tools/differ/differ.py

```diff
@@ -0,0 +1,167 @@
+
+variableMeasured   added  deleted  modified  same  total
+0   dcid:var1       1      0       0          0     1
+1   dcid:var2       0      2       1          1     4
+2   dcid:var3       0      0       1          0     1
+3   dcid:var4       0      2       0          0     2
+
+  '''
+
+  def __init__(self, groupby_columns, value_columns):
+    self.groupby_columns = groupby_columns.split(',')
+    self.value_columns = value_columns.split(',')
+    self.variable_column  =  self.groupby_columns[0]
+    self.place_column  =  self.groupby_columns[1]
+    self.time_column  =  self.groupby_columns[2]
+    self.diff_column = '_diff_result'
+
+  def __cleanup_data(self, df: pd.DataFrame):
+    for column in ['added', 'deleted', 'modified', 'same']:
+      df[column] = df[column] if column in df.columns else 0
+      df[column] = df[column].fillna(0).astype(int)
+
+  # Pro-rocesses two dataset files to identify changes.
+  def process_data(self, previous_df: pd.DataFrame, current_df: pd.DataFrame) -> pd.DataFrame:
+    cur_df_columns = current_df.columns.values.tolist()
+    self.groupby_columns = [i for i in self.groupby_columns if i in cur_df_columns]
+    self.value_columns = [i for i in self.value_columns if i in cur_df_columns]
+    df1 = previous_df.loc[:, self.groupby_columns + self.value_columns]
+    df2 = current_df.loc[:, self.groupby_columns + self.value_columns]
+    df1['_value_combined'] = df1[self.value_columns]\
+      .apply(lambda row: '_'.join(row.values.astype(str)), axis=1)
+    df2['_value_combined'] = df2[self.value_columns]\
+      .apply(lambda row: '_'.join(row.values.astype(str)), axis=1)
+    df1.drop(columns=self.value_columns, inplace=True)
+    df2.drop(columns=self.value_columns, inplace=True)
+    # Perform outer join operation to identify differences.
+    result  = pd.merge(df1, df2, on = self.groupby_columns, how='outer', indicator=self.diff_column)
+    result[self.diff_column] = result.apply(
+      lambda row: 'added' if row[self.diff_column] == 'right_only' \
+      else 'deleted' if row[self.diff_column] == 'left_only' \
+      else 'modified' if row['_value_combined_x'] != row['_value_combined_y'] \
+      else 'same', axis=1)
+    return result
+
+  # Performs point diff analysis to identify data point changes.
+  def point_analysis(self, in_data: pd.DataFrame) -> (pd.DataFrame, pd.DataFrame):
+    column_list = [self.variable_column, self.place_column, self.time_column, self.diff_column]
+    result = in_data.loc[:, column_list]
+    # summary = summary.groupby([variable,'result'], observed=True, as_index=False).agg(['count']).reset_index()
+    result = result.groupby([self.variable_column, self.diff_column], observed=True, as_index=False)[[self.place_column, self.time_column]].agg(lambda x: x.tolist())
+    result['size'] = result.apply(lambda row:len(row[self.place_column]), axis=1)
+    result[self.place_column] = result.apply(lambda row: row[self.place_column][0:SAMPLE_COUNT], axis=1)
```

Comment:
If the places are sorted, can use `random.sample(list, size)` from the module random

Reply:
Done